pdu/UDH.go

```diff
@@ -15,27 +17,32 @@ import (
 type UDH []InfoElement
 
 // UDHL return the length (number of octet) of the encoded UDH itself
-func (u UDH) UDHL() (l int) {
-	for i := range u {
-		l += len(u[i].Data)
+func (u UDH) UDHL() int {
+	length := 1
+	for index := range u {
+		length += 2
+		length += len(u[index].Data)
 	}
-	return l
+	return length
 }
 
 // MarshalBinary marshal UDH into bytes array
 // The first byte is UDHL
 // MarshalBinary preserve InformationElement order as they appears in the UDH
 func (u *UDH) MarshalBinary() (b []byte, err error) {
 	if len(*u) == 0 {
-		return nil, nil
+		return
 	}
-
-	b = []byte{byte(u.UDHL())}
-
-	for _, ie := range *u {
-		b = append(b, ie.Data...)
+	var buf bytes.Buffer
```

Comment:
How about preallocate with “udhl”:
 buf := bytes. NewBufffer(make([]byte, 0, c.UDHL())

Reply:
see https://golang.org/pkg/bytes/#NewBuffer

without this method signature.

Reply:
What do you means by without this method signature? `bytes.Buffer` will use underlying provided slice as storage:
```
buf := bytes. NewBufffer(make([]byte, 0, c.UDHL())
```

Reply:
```go
func NewBuffer(buf []byte) *Buffer
```

Reply:
Yeah, it's the signature, so what do you mean?

Reply:
I understand it become `bytes.NewBuffer([]byte, int, int)`.
Sorry, fixed.